java/code/src/com/suse/manager/maintenance/MaintenanceManager.java

```diff
@@ -695,6 +696,57 @@ public boolean isActionInMaintenanceWindow(Action action, MaintenanceSchedule sc
         return false;
     }
 
+    private Collection<CalendarComponent> getDateFromSchedule(MaintenanceSchedule ms) {
+        return getScheduleEventsAtDate(new Date(), ms, parseCalendar(ms.getCalendarOpt().get()));
+    }
+
+    /**
+     * Check if system is in maintenance mode
+     *
+     * @param server the server to check
+     * @return true when the action is inside of a maintenance window, otherwise false
+     */
+    public static boolean checkIfInMaintenanceMode(MinionServer server) {
+        if (server.getMaintenanceScheduleOpt().isEmpty()) {
+            return true;
+        }
+        else {
+            MaintenanceManager mm = MaintenanceManager.instance();
+            Collection<CalendarComponent> cc = mm.getDateFromSchedule(server.getMaintenanceScheduleOpt().get());
+            return cc.isEmpty();
+        }
+    }
+
+    /**
+     * Log the number of servers skipped and if debugging is enabled list the server names
+     *
+     * @param servers the list of servers to log
+     */
+    public static void logSkippedMinions(List<MinionServer> servers) {
+       log.warn("Skipping action for " + servers.size() + " minions.");
+       if (log.isDebugEnabled()) {
+          String serverNames = servers.stream()
+          .map(m -> m.getName())
+          .collect(Collectors.joining(","));
+          log.debug("Skipped minion ids: " + serverNames);
+       }
+    }
+
+    /**
+     * Given a list of minions, sorts by maintenance mode status, logs skipped minions
+     *
+     * @param minions servers to check
+     * @return List of minions in maintenance mode
+     */
+    public static List<Long> systemIdsMaintenanceMode(List<MinionServer> minions) {
+         Map<Boolean, List<MinionServer>> sepList = minions.stream()
+              .collect(Collectors.partitioningBy(m -> MaintenanceManager.checkIfInMaintenanceMode(m)));
```

Comment:
```suggestion
        Map<Optional<MaintenanceSchedule>, List<MinionServer>> minionsBySchedule = minions
                .stream().collect(Collectors.groupingBy(m -> m.getMaintenanceScheduleOpt()));

        Map<Boolean, List<MinionServer>> sepList = minionsBySchedule.entrySet()
                .stream().collect(Collectors.partitioningBy(mapEntry -> {
                    Optional<MaintenanceSchedule> optSchedule = mapEntry.getKey();
                    if (optSchedule.isEmpty()) {
                        return true;
                    }

                    MaintenanceManager mm = MaintenanceManager.instance();
                    return mm.getDateFromSchedule(optSchedule.get()).isEmpty();
                    }, Collectors.flatMapping(e -> e.getValue().stream(), Collectors.toList())));
```
Looks funny, but it should speedup things a lot as we do not check every single minion, but group them by maintenance schedules first and than perform the check only for every schedule.

The method `checkIfMaintenanceMode` is now unused. But it could be used in the `Server` object to perform this check for a single system.

Reply:
@mcalmer I like the idea, but I think the implementation could be simplified, maybe something like this would be more readable. The idea here is to compute which maintenance schedules have maintenance mode right now (this happens once per schedule (see the `distinct`)). Then we use this information for filtering minions:

```java
MaintenanceManager mm = MaintenanceManager.instance(); // (btw, the `systemIdsMaintenanceMode` would better not be static, so we could avoid this assignment)

// schedules that are in maintenance mode RIGHT NOW
Set<MaintenanceSchedule> schedulesInMaintMode = minions.stream()
        .flatMap(minion -> minion.getMaintenanceScheduleOpt().stream())
        .distinct()
        .filter(sched -> !mm.getDateFromSchedule(sched).isEmpty())
        .collect(Collectors.toSet());

List<Long> minionsInMaintMode = minions.stream()
        .filter(minion -> minion.getMaintenanceScheduleOpt()
                .map(sched -> schedulesInMaintMode.contains(sched)) // keep minions that have maintenance mode
                .orElse(true)) // or that have no maintenance schedule whatsoever
        .map(minion -> minion.getId())
        .collect(toList());

return minionsInMaintMode;

```

Reply:
We need both lists, the list which could execute the action and the one for the skipped systems as we want to log.

But it can also be achieved by creating the list of minions which can execute the action, and than use this:

```
minions.stream().filter(m -> !minionsInMaintMode.contains(m)).collect(toList())
```
for the logging